### ISSUE TITLE
mnemonic code enum

### ISSUE THREAD

**Issue by callumenator** [yazd/udis86-d#1]
What version of udis are you binding here? The ud_mnemonic_code enum you have in itab.d doesn't seem to map correctly to udis 1.7.2 (i.e. looking up the mnemonic codes doesn't work correctly). Copying the enum from itab.h source fixed this for me. 

Also, for some reason dmd doesn't like things like this:

```
writeln(ud_mnemonic_code.UD_Ijmp)
```

It fails with a recursive template expansion in EnumMembers. So I have also had to change the declaration to something like this:

```
alias uint ud_mnemonic_code;
enum : ud_mnemonic_code
{
...
}
```

One final thing - on windows I have to create the ud_t object on the heap:

```
auto ud_obj = new ud_t();
```

else when the program exits I get access violations (I am linking against the dll, after having converted the udis86.lib to OMF format). I assume you don't see this on linux, because everything is linked in a sane way, but I thought I would mention it, in case you knew the reason.

**Comment by yazd** [comment]
This version binds v1.7.2.

I did recheck the [header](https://gist.github.com/yazd/880449f78e3923f6e873) files on my system (Manjaro x86_64) and diffed the mnemonic enums and they matched exactly. Could it be that the headers are different on different platforms? That doesn't make much sense to me, but I will double check when I get home.

The reason for the template expansion error is that the enum is too large that DMD hits its limit of 500 'children' template instantiations when instantiating the `EnumMembers` template. I think it is worthy to change the code to what you have suggested. I'll do that when I get home.

Regarding the need to put `ud_t` object on the heap, I have no idea.

**Comment by callumenator** [comment]
Ah I see yes it must be a platform difference (I'm on x86), since that header is created by the build script. I can submit a PR with the X86 version if you need.

I also had to change the ud_type enum in types.d, your header seems to have a different format from https://github.com/vmt/udis86/blob/master/libudis86/types.h, I had to copy that one to make it work.

**Comment by yazd** [comment]
If you can submit a PR for x86, that would be great.

And, ughh, you are right regarding `ud_type`. I don't understand how I got that wrong! Fixed on master.

**Comment by callumenator** [comment]
Looking further, I can't see a platform difference in the build system. The build system calls a python script to generate itab.h, the python script uses udis86/docs/x86/optable.xml. My local build generates both x86 and x86_64 dlls from the same itab.h.

Could you try disassembling just [0x37], and check:

```
assert(ud_insn_mnemonic(&ud_obj) == UD_Iaaa);
```

That might clear it up.

**Comment by yazd** [comment]
The assert passes successfully.
I have looked online and I found different versions of itab.h:
https://cgit.haiku-os.org/haiku/plain/headers/libs/udis86/libudis86/itab.h
https://code.google.com/p/opendbg/source/browse/branches/minidbg/inc/libudis86/itab.h?r=126
https://github.com/radare/radare2/blob/master/shlr/udis86/itab.h

And I found that the other languages bindings each bind to only one header file 'version' with no regard to the other 'versions':
https://github.com/jroimartin/udis86
https://github.com/sophsec/ffi-udis86/blob/master/lib/ffi/udis86/types.rb
https://github.com/kmcallister/hdis86/blob/master/Hdis86/Internal/Opcode.hs (this binding has the c library under it, so it is not affected)

**Comment by callumenator** [comment]
Yeah I was hoping the mnemonic codes were constant as far as the library is concerned, but it looks like the library needs to use the same itab.h (.d) it was built against. Not sure how a binding can handle this well. It would be fine if compilation failed or something, but instead it introduces subtle errors in mnemonic lookup.
